lib/aur-repo

```diff
@@ -0,0 +1,132 @@
+#!/bin/bash
+# aur-repo - manage local repositories
+readonly argv0=repo
+readonly PS4='+(${BASH_SOURCE}:${LINENO}): ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
+
+# default arguments
+vercmp_args=()
+
+conf_file_repo() {
+    awk -F'= ' '
+        $1 ~ /^\[.+\]$/ {
+            repo = substr($1, 2, length($1)-2)
+        }
+        $1 ~ /^Server/ && $2 ~ /^file:/ {
+            printf("%s\n%s\n", repo, $2)
+        }'
+}
+
+db_namever() {
+    awk '/%NAME%/ {
+        getline
+        printf("%s\t", $1)
+    }
+    /%VERSION%/ {
+        getline
+        printf("%s\n", $1)
+    }
+    END {
+        if (!NR)
+            printf("warning: empty database\n") > "/dev/stderr"
+    }'
+}
+
+check_sync() {
+    # https://github.com/andrewgregory/pacutils/issues/22
+    pacconf --repo="$1" >/dev/null || return 1
+}
+
+parse_sync() {
+    # https://git.archlinux.org/pacman.git/commit/?id=ab3d8478
+    pacsift --exact --null --repo "$1" <&- \
+        | xargs -0r pacman -Sddp --print-format "$2"
+}
+
+usage() {
+    plain "usage: $argv0 [-aSlu] [-d repo] [-r root]"
+    exit 1
+}
+
+source /usr/share/makepkg/util/message.sh
+source /usr/share/makepkg/util/parseopts.sh
+
+if [[ -t 2 && ! -o xtrace ]]; then
+    colorize
+fi
+
+## option parsing
+opt_short='d:o:r:ahlSu'
+opt_long=('all' 'database:' 'list' 'root:' 'pacman' 'upgrades'
+          'repo-list' 'status-file:' 'help')
+
+if ! parseopts "$opt_short" "${opt_long[@]}" -- "$@"; then
+    usage
+fi
+set -- "${OPTRET[@]}"
+
+unset mode db_name db_root status_file
+while true; do
+    case $1 in
+        -a|--all)      vercmp_args+=(--all) ;;
+        -d|--database) shift; db_name=$1 ;;
+        -r|--root)     shift; db_root=$1 ;;
+        --status-file) shift; status_file=$1 ;;
+        -l|--list)     mode=list_packages ;;
+        -S|--pacman)   mode=list_pacman ;;
+        -u|--upgrades) mode=list_upgrades ;;
+        --repo-list)   mode=repo_list ;;
+        -h|--help) usage ;;
+        --) shift; break ;;
+    esac
+    shift
+done
+unset opt_short opt_long OPTRET
+
+case $mode in
+    repo_list) pacconf | conf_file_repo
+               exit ;;
+esac
+
+if [[ ${db_name=$AUR_REPO} ]]; then
+    if ! [[ ${db_root=$AUR_DBROOT} ]]; then
+        db_root=$(pacconf --single --repo="$db_name" Server) || exit 1
+        db_root=${db_root#*://}
+    fi
```

Comment:
Maybe add a test that db_root is a 'file://' and not an 'http://'?

Reply:
This should cover all cases:
```
    if ! [[ ${db_root=$AUR_DBROOT} ]]; then
        db_root=$(pacconf --single --repo="$db_name" Server)
        db_root=${db_root#file://}

        # check if variable was set correctly
        if ! [[ $db_root ]]; then
            error "$argv0: Server directive not found"
            exit 1
        fi

        # check if variable points to a local path (file://)
        if [[ $db_root == *://* ]]; then
            error "$argv0: Server: not a local directory"
            exit 1
        fi

        # check if local path is a directory
        if ! [[ -d $db_root ]]; then
            error "argv0: Server: not a directory"
            exit 20
        fi
    fi
```